### PR TITLE
Unary Ops in FP8 Windowed Einsums

### PR DESCRIPTION
**Pull request by philipphack**
Adds support for unary ops between dequantization and windowed einsum loop.

### CODE REVIEW

**Comment by cheshire** [comment]
Could we somehow go back to the drawing board and figure out how to do this without excessive pattern matching?

**Comment by philipphack** [comment]
@cheshire the windowed einsum handler effectively pattern matches and transforms windowed einsum loops created by the SPMD dot handler. The alternative would be the direct modification of the dot handler which is already relatively complex.

File: xla/service/gpu/transforms/windowed_einsum_handler.cc [hunk 1]
```diff
@@ -56,11 +56,14 @@ namespace m = match;
 // and type conversions of FP8 operands into the bodies of their while loops,
 // i.e. rewrites
 //
-//   inputs --> dequant --> while loop {collective-permute/dot/etc}
+//   inputs --> dequant --> (unary) --> while loop {collective-permute/dot/etc}
 //
 // into
 //
-//   inputs --> while loop {dequant --> collective-permute/dot/etc}.
+//   inputs --> while loop {(unary) --> dequant --> collective-permute/dot/etc}.
```
Comment (reedwm) [inline xla/service/gpu/transforms/windowed_einsum_handler.cc:63]: The unary ops are stlil outside the while loop. So this line should be

```
//   inputs --> (unary) --> while loop {dequant --> collective-permute/dot/etc}.
```

File: xla/service/gpu/transforms/windowed_einsum_handler_test.cc [hunk 2]
```diff
@@ -634,127 +634,136 @@ CHECK: ROOT {{.*}} = bf16[1,4,1,1,2048,8192]{5,4,3,2,1,0} reshape(bf16[1,4,1,204
...
+; CHECK-NEXT:    [[C2:%[^ ]+]] = u32[] constant(2)
+; CHECK-NEXT:    [[NEXT_LOOP_COUNTER:%[^ ]+]] = u32[] add([[LOOP_COUNTER]], [[C2]])
+; CHECK-NEXT:    [[TUPLE:%[^ ]+]] = (f8e4m3fn[2,512,24576]{2,1,0}, f8e4m3fn[24576,24576]{1,0}, f32[2,2048,24576]{2,1,0}, f32[2,2048,24576]{2,1,0}, u32[], /*index=5*/f32[], f32[]) tuple([[PERMUTED_LHS1]], [[RHS]], [[UPDATED_DOT_OUTPUT1]], [[PASS_THROUGH]], [[NEXT_LOOP_COUNTER]], /*index=5*/[[SCALE_LHS]], [[SCALE_RHS]])
+; CHECK-LABEL: ENTRY %main
+; CHECK:         [[LHS:%[^ ]+]] = f8e4m3fn[2,512,24576]{2,1,0} parameter(0), sharding={devices=[1,4,1]<=[4]}
+; CHECK-NEXT:    [[RHS:%[^ ]+]] = f8e4m3fn[1536,24576]{1,0} parameter(1), sharding={devices=[1,4]<=[4]}
+; CHECK-NEXT:    [[RHS_BCAST:%[^ ]+]] = f8e4m3fn[16,1536,24576]{2,1,0} broadcast([[RHS]]), dimensions={1,2}
+; CHECK-NEXT:    [[RHS_RESHAPED:%[^ ]+]] = f8e4m3fn[24576,24576]{1,0} reshape([[RHS_BCAST]])
+; CHECK:         [[WHILE0:%[^ ]+]] = (f8e4m3fn[2,512,24576]{2,1,0}, f8e4m3fn[24576,24576]{1,0}, f32[2,2048,24576]{2,1,0}, f32[2,2048,24576]{2,1,0}, u32[], /*index=5*/f32[], f32[]) while([[TUPLE0:%[^ ]+]]),
```
Comment (reedwm) [inline xla/service/gpu/transforms/windowed_einsum_handler_test.cc:764]: Also check the RHS_RESHAPED is actually passed to the while-loop. If there were a bug that mistakenly passed the original RHS to the while-loop, no test would catch it right now.

Also I would add an end to end test.

**Review by reedwm — CHANGES_REQUESTED** [review]

File: xla/tests/collective_ops_e2e_test.cc [hunk 3]
```diff
@@ -808,46 +808,62 @@ ENTRY main.12 {
...
+  scale_rhs = bf16[] parameter(3)
+  scale_lhs_bcast = bf16[2,16,48]{2,1,0} broadcast(scale_rhs), dimensions={}
+  scale_rhs_bcast = bf16[2,24,192]{2,1,0} broadcast(scale_lhs), dimensions={}
+  lhs_bf16 = bf16[2,16,48]{2,1,0} convert(lhs)
+  rhs_bf16 = bf16[2,24,192]{2,1,0} convert(rhs)
+  lhs_scaled = bf16[2,16,48]{2,1,0} multiply(scale_lhs_bcast, lhs_bf16)
+  rhs_scaled = bf16[2,24,192]{2,1,0} multiply(scale_rhs_bcast, rhs_bf16)
+  rhs_reshaped = bf16[48,192]{1,0} reshape(rhs_scaled)
+  dot = bf16[2,16,192]{2,1,0} dot(lhs_scaled, rhs_reshaped), lhs_contracting_dims={2}, rhs_contracting_dims={0}
+  ROOT custom-call = bf16[2,16,192]{2,1,0} custom-call(dot), custom_call_target="Sharding", sharding={devices=[1,1,4]<=[4]}
```
Comment (reedwm) [inline xla/tests/collective_ops_e2e_test.cc:865]: Just put the sharding annotation on the dot itself, by adding to the end: `, sharding={devices=[1,1,4]<=[4]}`

File: xla/tests/collective_ops_e2e_test.cc [hunk 4]
```diff
@@ -808,46 +808,62 @@ ENTRY main.12 {
   CollectiveOpsCompareWindowedNonWindowed(kModuleReplicatedStr);
 }
 
+TEST_F(CollectiveOpsTestE2EWindowedNonWindowed, WindowedEinsumE2EAllGatherF8) {
+  absl::string_view kModuleReplicatedStr = R"(
+HloModule pjit__unnamed_wrapped_function_, entry_computation_layout={(f8e4m3fn[2,16,48]{2,1,0}, f8e4m3fn[48,192]{1,0}, bf16[], bf16[])->bf16[2,16,192]{2,1,0}}, allow_spmd_sharding_propagation_to_parameters={false,false,false,false}, num_partitions=4
+
+ENTRY main {
+  lhs = f8e4m3fn[2,16,48]{2,1,0} parameter(0), sharding={devices=[1,4,1]<=[4]}
```
Comment (reedwm) [inline xla/tests/collective_ops_e2e_test.cc:816]: Why have a two non-contracting dims? It's simply if there's just one.

File: xla/tests/collective_ops_e2e_test.cc [hunk 5]
```diff
@@ -872,24 +888,61 @@ TEST_F(CollectiveOpsTestE2EWindowedNonWindowed,
 HloModule windowed_einsum_e2e_all_gather_multi_consumer_f8, entry_computation_layout={(f8e4m3fn[2,16,48]{2,1,0}, f8e4m3fn[48,192]{1,0}, f8e4m3fn[48,192]{1,0}, bf16[], bf16[], bf16[])->bf16[2,16,192]{2,1,0}}, allow_spmd_sharding_propagation_to_parameters={false,false,false,false}, num_partitions=4
 
 ENTRY main {
-  rhs = f8e4m3fn[2,16,48]{2,1,0} parameter(0), sharding={devices=[1,4,1]<=[4]}
-  lhs0 = f8e4m3fn[48,192]{1,0} parameter(1), sharding={devices=[1,4]<=[4]}
+  lhs = f8e4m3fn[2,16,48]{2,1,0} parameter(0), sharding={devices=[1,4,1]<=[4]}
```
Comment (reedwm) [inline xla/tests/collective_ops_e2e_test.cc:891]: Why change this test?

**Review by reedwm — CHANGES_REQUESTED** [review]

**Review by reedwm — APPROVED** [review]

**Pull request closed without merging**